gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/policy/PartitionAwareHiveRegistrationPolicy.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.hive.policy;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.hive.HivePartition;
+import org.apache.gobblin.hive.HiveRegistrationUnit;
+import org.apache.gobblin.hive.HiveTable;
+
+
+/**
+ * This is the partition aware implementation of {@link HiveRegistrationPolicyBase}.
+ * You can specify hive.partition.regex where the first match is the table location and the
+ * rest match are the partitions.
+ * The partition names can be specified with the hive.table.partition.keys property.
+ * The order of the partition names should match the order of regexp matches in the hive.partition.regexp expression.
+ *
+ * For example in the case of path: s3://testbucket/myawesomlogs/compacted/dt=20170101/hr=22/
+ * The hive.partition.regexp would look like: hive.partition.regex=(s3://testbucket/myawesomelogs/compacted/)dt=(.*)/hr=(.*)
```

Comment:
I wonder if there are no recognizable key in the partition identifier (Like what you have here `dt`, `hr`), what should we do about it ?